File: benchmarks/79-shgo/shgo.py
#!/usr/bin/env python3
"""Experiment 79 -- SHGO global optimization (scipy column).

Same 8 problems as ``shgo.m``, same labels, same order, solved with
scipy.optimize.shgo using the SAME sampling_method and sample count (n) on both
sides, so the race is SHGO-vs-shgo (both build a sampled complex, find the
minimizer pool, and multi-start a local minimizer from it). These are the
standard bounded multimodal benchmarks -- Himmelblau, Branin, six-hump camel,
Cross-in-tray, Shubert, Rastrigin, Ackley, Styblinski-Tang -- where the pool +
homology construction is designed to shine. The check is the objective at the
global optimum rounded to 10^6, which both systems reach and polish to
agreement.
"""

import sys, os; sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

import numpy as np
from scipy.optimize import shgo

from harness import bench, check, require

require(["scipy.optimize:shgo"])


def _solve(fun, bounds, n, method, iters=1):
    return shgo(fun, bounds, n=n, iters=iters, sampling_method=method)


# ---- 2-D benchmarks (simplicial) ----
def _himmelblau(v):
    x, y = v
    return (x * x + y - 11.0) ** 2 + (x + y * y - 7.0) ** 2

def _branin(v):
    x, y = v
    return ((y - 5.1 / (4 * np.pi ** 2) * x * x + 5 / np.pi * x - 6.0) ** 2
            + 10 * (1 - 1 / (8 * np.pi)) * np.cos(x) + 10)

def _camel(v):
    x, y = v
    return (4 - 2.1 * x * x + x ** 4 / 3) * x * x + x * y + (-4 + 4 * y * y) * y * y

def _cross_in_tray(v):
    x, y = v
    return -0.0001 * (abs(np.sin(x) * np.sin(y)
                          * np.exp(abs(100 - np.sqrt(x * x + y * y) / np.pi))) + 1) ** 0.1

def _shubert(v):
    x, y = v
    sx = sum(j * np.cos((j + 1) * x + j) for j in range(1, 6))
    sy = sum(j * np.cos((j + 1) * y + j) for j in range(1, 6))
    return sx * sy


# ---- n-D benchmarks ----
def _rastrigin(v):
    v = np.asarray(v)
    return 10.0 * v.size + np.sum(v * v - 10.0 * np.cos(2 * np.pi * v))

def _ackley(v):
    v = np.asarray(v)
    n = v.size
    return (-20.0 * np.exp(-0.2 * np.sqrt(np.sum(v * v) / n))
            - np.exp(np.sum(np.cos(2 * np.pi * v)) / n) + 20.0 + np.e)

def _styblinski(v):
    v = np.asarray(v)
    return float(np.sum(v ** 4 - 16 * v * v + 5 * v) / 2.0)


_c01 = lambda: _solve(_himmelblau,   [(-5, 5), (-5, 5)],          100, "simplicial")
_c02 = lambda: _solve(_branin,       [(-5, 10), (0, 15)],         100, "simplicial")
_c03 = lambda: _solve(_camel,        [(-3, 3), (-2, 2)],          200, "simplicial")
_c04 = lambda: _solve(_cross_in_tray,[(-10, 10), (-10, 10)],      500, "simplicial")
_c05 = lambda: _solve(_shubert,      [(-10, 10), (-10, 10)],      600, "simplicial")
_c06 = lambda: _solve(_rastrigin,    [(-5.12, 5.12)] * 3,         200, "simplicial")
_c07 = lambda: _solve(_ackley,       [(-5, 5)] * 4,               200, "simplicial")
_c08 = lambda: _solve(_styblinski,   [(-5, 5)] * 4,               256, "sobol")

bench("01 Himmelblau simplicial", _c01)
check("01 Himmelblau simplicial", int(np.floor(1e6 * _c01().fun + 0.5)))

bench("02 Branin simplicial", _c02)
check("02 Branin simplicial", int(np.floor(1e6 * _c02().fun + 0.5)))

bench("03 Six-hump camel simplicial", _c03)
check("03 Six-hump camel simplicial", int(np.floor(1e6 * _c03().fun + 0.5)))

bench("04 Cross-in-tray simplicial", _c04)
check("04 Cross-in-tray simplicial", int(np.floor(1e6 * _c04().fun + 0.5)))

bench("05 Shubert simplicial", _c05)
check("05 Shubert simplicial", int(np.floor(1e6 * _c05().fun + 0.5)))

bench("06 Rastrigin 3D simplicial", _c06)
check("06 Rastrigin 3D simplicial", int(np.floor(1e6 * _c06().fun + 0.5)))

bench("07 Ackley 4D simplicial", _c07)
check("07 Ackley 4D simplicial", int(np.floor(1e6 * _c07().fun + 0.5)))

bench("08 Styblinski-Tang 4D sobol", _c08)
check("08 Styblinski-Tang 4D sobol", int(np.floor(1e6 * _c08().fun + 0.5)))

File: benchmarks/79-shgo/shgo.m
(* Experiment 79 -- SHGO (Simplicial Homology Global Optimization) global
   optimization (NMinimize, Method -> {"SHGO", ...}).

   MEASURES src/numerical_calculus/findmin.c's nm_shgo -- the method samples the
   bounded box (the exact Kuhn-triangulation 1-skeleton for "Simplicial", a
   k-nearest-neighbour graph over a Sobol'/Halton low-discrepancy set otherwise),
   identifies the minimizer pool (vertices better than every graph neighbour),
   and starts one local polish from each pool vertex -- against
   scipy.optimize.shgo. Same problems, same labels, same order, same
   sampling_method and sample count on both sides, so the race is
   SHGO-vs-shgo. These are the standard bounded multimodal benchmarks where the
   pool + homology construction shines: Himmelblau, Branin, six-hump camel,
   Cross-in-tray, Shubert, Rastrigin, Ackley, Styblinski-Tang. The check is the
   objective at the global optimum rounded to 10^6, which both systems reach and
   polish to agreement. Objectives over several variables are built with
   Sum/Table over fresh global symbols and passed through Evaluate[] past
   NMinimize's argument handling. *)

Get["../harness.m"];

require[{"NMinimize"}];

(* n-D objectives over fresh global symbols. *)
rastrigin[n_] := With[{v = Table[Symbol["rv" <> ToString[i]], {i, 1, n}]},
                      10 n + Sum[v[[i]]^2 - 10 Cos[2 Pi v[[i]]], {i, 1, n}]];
rastriginV[n_] := Table[Symbol["rv" <> ToString[i]], {i, 1, n}];
rastriginB[n_] := And @@ Table[-5.12 <= Symbol["rv" <> ToString[i]] <= 5.12, {i, 1, n}];

ackley[n_] := With[{v = Table[Symbol["av" <> ToString[i]], {i, 1, n}]},
                   -20 Exp[-0.2 Sqrt[Sum[v[[i]]^2, {i, 1, n}]/n]]
                   - Exp[Sum[Cos[2 Pi v[[i]]], {i, 1, n}]/n] + 20 + E];
ackleyV[n_] := Table[Symbol["av" <> ToString[i]], {i, 1, n}];
ackleyB[n_] := And @@ Table[-5 <= Symbol["av" <> ToString[i]] <= 5, {i, 1, n}];

styb[n_] := With[{v = Table[Symbol["sv" <> ToString[i]], {i, 1, n}]},
                 Sum[v[[i]]^4 - 16 v[[i]]^2 + 5 v[[i]], {i, 1, n}]/2];
stybV[n_] := Table[Symbol["sv" <> ToString[i]], {i, 1, n}];
stybB[n_] := And @@ Table[-5 <= Symbol["sv" <> ToString[i]] <= 5, {i, 1, n}];

(* Precompute out of the timed region. *)
ra3 = rastrigin[3]; ra3v = rastriginV[3]; ra3b = rastriginB[3];
ac4 = ackley[4];    ac4v = ackleyV[4];    ac4b = ackleyB[4];
st4 = styb[4];      st4v = stybV[4];      st4b = stybB[4];

(* 2-D benchmarks, simplicial sampling. *)
bench["01 Himmelblau simplicial", NMinimize[{(x^2+y-11)^2+(x+y^2-7)^2, -5<=x<=5 && -5<=y<=5}, {x,y}, Method->{"SHGO","SamplingMethod"->"Simplicial","SearchPoints"->100}];];
check["01 Himmelblau simplicial", Round[10^6 First[NMinimize[{(x^2+y-11)^2+(x+y^2-7)^2, -5<=x<=5 && -5<=y<=5}, {x,y}, Method->{"SHGO","SamplingMethod"->"Simplicial","SearchPoints"->100}]]]];

bench["02 Branin simplicial", NMinimize[{(y-5.1/(4 Pi^2) x^2+5/Pi x-6)^2+10(1-1/(8 Pi))Cos[x]+10, -5<=x<=10 && 0<=y<=15}, {x,y}, Method->{"SHGO","SamplingMethod"->"Simplicial","SearchPoints"->100}];];
check["02 Branin simplicial", Round[10^6 First[NMinimize[{(y-5.1/(4 Pi^2) x^2+5/Pi x-6)^2+10(1-1/(8 Pi))Cos[x]+10, -5<=x<=10 && 0<=y<=15}, {x,y}, Method->{"SHGO","SamplingMethod"->"Simplicial","SearchPoints"->100}]]]];

bench["03 Six-hump camel simplicial", NMinimize[{(4-2.1 x^2+x^4/3)x^2+x y+(-4+4 y^2)y^2, -3<=x<=3 && -2<=y<=2}, {x,y}, Method->{"SHGO","SamplingMethod"->"Simplicial","SearchPoints"->200}];];
check["03 Six-hump camel simplicial", Round[10^6 First[NMinimize[{(4-2.1 x^2+x^4/3)x^2+x y+(-4+4 y^2)y^2, -3<=x<=3 && -2<=y<=2}, {x,y}, Method->{"SHGO","SamplingMethod"->"Simplicial","SearchPoints"->200}]]]];

bench["04 Cross-in-tray simplicial", NMinimize[{-0.0001(Abs[Sin[x]Sin[y]Exp[Abs[100-Sqrt[x^2+y^2]/Pi]]]+1)^0.1, -10<=x<=10 && -10<=y<=10}, {x,y}, Method->{"SHGO","SamplingMethod"->"Simplicial","SearchPoints"->500}];];
check["04 Cross-in-tray simplicial", Round[10^6 First[NMinimize[{-0.0001(Abs[Sin[x]Sin[y]Exp[Abs[100-Sqrt[x^2+y^2]/Pi]]]+1)^0.1, -10<=x<=10 && -10<=y<=10}, {x,y}, Method->{"SHGO","SamplingMethod"->"Simplicial","SearchPoints"->500}]]]];

bench["05 Shubert simplicial", NMinimize[{(Sum[j Cos[(j+1)x+j],{j,1,5}])(Sum[j Cos[(j+1)y+j],{j,1,5}]), -10<=x<=10 && -10<=y<=10}, {x,y}, Method->{"SHGO","SamplingMethod"->"Simplicial","SearchPoints"->600}];];
check["05 Shubert simplicial", Round[10^6 First[NMinimize[{(Sum[j Cos[(j+1)x+j],{j,1,5}])(Sum[j Cos[(j+1)y+j],{j,1,5}]), -10<=x<=10 && -10<=y<=10}, {x,y}, Method->{"SHGO","SamplingMethod"->"Simplicial","SearchPoints"->600}]]]];

(* n-D benchmarks, Sobol'/Simplicial. *)
bench["06 Rastrigin 3D simplicial", NMinimize[{Evaluate[ra3], Evaluate[ra3b]}, Evaluate[ra3v], Method->{"SHGO","SamplingMethod"->"Simplicial","SearchPoints"->200}];];
check["06 Rastrigin 3D simplicial", Round[10^6 First[NMinimize[{Evaluate[ra3], Evaluate[ra3b]}, Evaluate[ra3v], Method->{"SHGO","SamplingMethod"->"Simplicial","SearchPoints"->200}]]]];

bench["07 Ackley 4D simplicial", NMinimize[{Evaluate[ac4], Evaluate[ac4b]}, Evaluate[ac4v], Method->{"SHGO","SamplingMethod"->"Simplicial","SearchPoints"->200}];];
check["07 Ackley 4D simplicial", Round[10^6 First[NMinimize[{Evaluate[ac4], Evaluate[ac4b]}, Evaluate[ac4v], Method->{"SHGO","SamplingMethod"->"Simplicial","SearchPoints"->200}]]]];

bench["08 Styblinski-Tang 4D sobol", NMinimize[{Evaluate[st4], Evaluate[st4b]}, Evaluate[st4v], Method->{"SHGO","SamplingMethod"->"Sobol","SearchPoints"->256}];];
check["08 Styblinski-Tang 4D sobol", Round[10^6 First[NMinimize[{Evaluate[st4], Evaluate[st4b]}, Evaluate[st4v], Method->{"SHGO","SamplingMethod"->"Sobol","SearchPoints"->256}]]]];
